test/activerecord_spanner_mock_server/spanner_active_record_with_mock_server_test.rb

```diff
@@ -0,0 +1,280 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+require_relative "../mock_server/spanner_mock_server"
+require_relative "../test_helper"
+require_relative "models/singer"
+require_relative "models/album"
+
+require "securerandom"
+
+class SpannerActiveRecordMockServerTest < Minitest::Test
+  def setup
+    super
+    @server = GRPC::RpcServer.new
+    @port = @server.add_http2_port "localhost:0", :this_port_is_insecure
+    @mock = SpannerMockServer.new
+    @server.handle(@mock)
+    # Run the server in a separate thread
+    @server_thread = Thread.new do
+      @server.run
+    end
+    # Register INFORMATION_SCHEMA queries on the mock server.
+    register_select_tables_result
+    register_singers_columns_result
+    register_singers_primary_key_result
+    register_singer_index_columns_result
+    # Connect ActiveRecord to the mock server
+    ActiveRecord::Base.establish_connection(
+      adapter: "spanner",
+      emulator_host: "localhost:#{@port}",
+      project: "test-project",
+      instance: "test-instance",
+      database: "testdb"
+    )
+    ActiveRecord::Base.logger = nil
+  end
+
+  def teardown
+    super
+    ActiveRecord::Base.connection_pool.disconnect!
+    @server.stop
+    @server_thread.exit
+  end
+
+  def test_selects_all_singers_without_transaction
+    sql = "SELECT `singers`.* FROM `singers`"
+    @mock.put_statement_result sql, create_random_singers_result(4)
+    Singer.all.each do |singer|
+      assert singer.id != nil
+      assert singer.first_name != nil
+      assert singer.last_name != nil
+    end
+    # None of the requests should use a (read-only) transaction.
+    select_requests = @mock.requests.select { |req| req.is_a? V1::ExecuteSqlRequest }
+    select_requests.each do |request|
+      assert_nil request.transaction
+    end
+    # Executing a simple query should not initiate any transactions.
+    begin_transaction_requests = @mock.requests.select { |req| req.is_a? V1::BeginTransactionRequest }
+    assert_empty begin_transaction_requests
+  end
+
+  def test_selects_one_singer_without_transaction
+    sql = "SELECT `singers`.* FROM `singers` WHERE `singers`.`id` = 1 LIMIT 1"
+    @mock.put_statement_result sql, create_random_singers_result(1)
+    singer = Singer.find_by id: 1
+
+    refute_nil singer
+    refute_nil singer.first_name
+    refute_nil singer.last_name
+    # None of the requests should use a (read-only) transaction.
```

Comment:
Queries outside of a transaction block should use a default single-use read-only transaction. This will be handled by the backend when a query is received without a transaction selector.